packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/ModelProvider.cs

```diff
@@ -65,14 +65,29 @@ public ModelProvider(InputModelType inputModel)
 
         private IReadOnlyList<ModelProvider>? _derivedModels;
         public IReadOnlyList<ModelProvider> DerivedModels => _derivedModels ??= BuildDerivedModels();
+        public ModelProvider? BaseModelProvider
+            => _baseModelProvider ??= (_baseTypeProvider?.Value is ModelProvider baseModelProvider ? baseModelProvider : null);
 
         private IReadOnlyList<ModelProvider> BuildDerivedModels()
         {
-            var derivedModels = new List<ModelProvider>(_inputModel.DiscriminatedSubtypes.Count);
+            var discriminatedSubTypes = new HashSet<ModelProvider>();
+            var derivedModels = new List<ModelProvider>(_inputModel.DiscriminatedSubtypes.Count + _inputModel.DerivedModels.Count);
             foreach (var subtype in _inputModel.DiscriminatedSubtypes)
             {
                 var model = CodeModelPlugin.Instance.TypeFactory.CreateModel(subtype.Value);
                 if (model != null)
+                {
+                    discriminatedSubTypes.Add(model);
+                }
+            }
+
+            derivedModels.AddRange(discriminatedSubTypes);
+
+            // add any derived models that are not in the discriminated subtypes
+            foreach (var derivedModel in _inputModel.DerivedModels)
+            {
+                var model = CodeModelPlugin.Instance.TypeFactory.CreateModel(derivedModel);
+                if (model != null && !discriminatedSubTypes.Contains(model))
```

Comment:
Could we just add directly to the HashSet for both derived and discriminated subtypes so it will handle duplicates for us?